testsuite/classlibrary/TestEnv.sc

```diff
@@ -26,6 +35,85 @@ TestEnv : UnitTest {
 			"Exponential Env.cutoff should be same as linear Env.cutoff.linexp");
 	}
 
+	test_noff_bug334 {
+		var synth, n_off_resp, n_end_resp, passed,
+		cond = Condition.new,
+		cleanup = {
+			n_off_resp.free;
+			n_end_resp.free;
+			server.quit;
+		};
+		this.bootServer(server);
+		synth = {
+			// doneAction: 1 should reach end *after* doneAction: 2 envelope ends
+			// correct behavior is that '/n_off' is never sent
+			EnvGen.kr(Env([0, 1], [0.2]), gate: 0, doneAction: 1);
+			EnvGen.kr(Env([0, 1], [0.1]), doneAction: 2);
+		}.play(server);
+		n_off_resp = OSCFunc({
+			cleanup.value;
+			passed = false;
+			cond.unhang;
+		}, '/n_off', server.addr, argTemplate: [synth.nodeID]);
+		// responsible for cleaning up this test
+		// n_off_resp should NOT fire! so we need another for the stop condition
+		n_end_resp = OSCFunc({
+			cleanup.value;
+			passed = true;
+			cond.unhang;
+		}, '/n_end', server.addr, argTemplate: [synth.nodeID]);
+		cond.hang;
+		this.assert(passed, "EnvGen: Initial gate == 0 with doneAction == 1 should not send /n_off");
+	}
+
+	// has failed in supernova
+	// methodology: synth should play for 0.1 sec if it's working
+	// so we will send a trigger just before the envelope ends
+	// if the bug occurs, the synth will end almost immediately -- no trigger
+	test_nongated_bug3094 {
+		var n_end_resp, tr_resp, synth,
+		program = [\scsynth], failure,
+		cleanup = {
+			tr_resp.free;
+			n_end_resp.free;
+			server.quit;
+			Server.scsynth;
+		},
+		cond = Condition.new;
+		// we need to test supernova.
+		// In unixy systems, we can try to get the executable path
+		// Windows, I don't know
+		if(thisProcess.platform.name != "windows" and: {
+			"which supernova".unixCmdGetStdOut.size > 0
+		}) {
+			program = program ++ [\supernova];
+		};
+		program.do { |p|
+			var gotTrigger = false;
+			Server.perform(p);
+			this.bootServer(server);
+			synth = {
+				var gate = Impulse.kr(0);
+				SendTrig.kr(TDelay.kr(gate, 0.09), 0, 1);
+				EnvGen.kr(Env([0, 1], [0.1]), gate, doneAction: 2);
+			}.play(server);
+			tr_resp = OSCFunc({ |msg|
+				gotTrigger = true;
+			}, '/tr', server.addr, argTemplate: [synth.nodeID]);
+			n_end_resp = OSCFunc({ |msg|
+				cleanup.value;
+				cond.unhang;
+			}, '/n_end', server.addr, argTemplate: [synth.nodeID]);
+			cond.hang;
+			if(gotTrigger.not) { failure = failure.add(p) };
+		};
+		this.assert(failure.size == 0, "EnvGen: Closing gate should not affect non-gated envelope %".format(
```

Comment:
When I wrote this, I was trying to follow the unit testing guideline to have one assert per test method. Since then, I had received multiple review comments saying that it's better to have multiple asserts, for clarity. So if I were writing it now, I would move the assert into the loop and just assert `gotTrigger`, and include the server program name in the message.

But maybe the current way is clear enough? Again, I'm just echoing review comments I'd received previously.